src/client.rs

```diff
@@ -3,76 +3,117 @@
 // Licensed under the MIT License, <LICENSE or http://opensource.org/licenses/MIT>.
 // This file may not be copied, modified, or distributed except according to those terms.
 
-use Packet;
-use futures::{Async, BoxFuture};
-use futures::stream::Empty;
+use WireError;
+use bincode::serde::DeserializeError;
+use framed::Framed;
+use futures::{self, Async, Future};
+use serde::{Deserialize, Serialize};
 use std::fmt;
 use std::io;
-use tokio_proto::pipeline;
+use tokio_core::net::TcpStream;
+use tokio_core::reactor;
+use tokio_proto::easy::{EasyClient, EasyResponse, multiplex};
 use tokio_service::Service;
-use util::Never;
 
 /// A client `Service` that writes and reads bytes.
 ///
 /// Typically, this would be combined with a serialization pre-processing step
 /// and a deserialization post-processing step.
-#[derive(Clone)]
-pub struct Client {
-    inner: pipeline::Client<Packet, Vec<u8>, Empty<Never, io::Error>, io::Error>,
+pub struct Client<Req, Resp, E> {
+    inner: EasyClient<Req, WireResponse<Resp, E>>,
 }
 
-impl Service for Client {
-    type Request = Packet;
-    type Response = Vec<u8>;
+type WireResponse<Resp, E> = Result<Result<Resp, WireError<E>>, DeserializeError>;
+type ResponseFuture<Resp, E> = futures::Map<EasyResponse<WireResponse<Resp, E>>,
+                                            fn(WireResponse<Resp, E>) -> Result<Resp, ::Error<E>>>;
+
+impl<Req, Resp, E> Service for Client<Req, Resp, E>
+    where Req: Send + 'static,
+          Resp: Send + 'static,
+          E: Send + 'static
+{
+    type Request = Req;
+    type Response = Result<Resp, ::Error<E>>;
     type Error = io::Error;
-    type Future = BoxFuture<Vec<u8>, io::Error>;
+    type Future = ResponseFuture<Resp, E>;
 
     fn poll_ready(&self) -> Async<()> {
-        Async::Ready(())
+        self.inner.poll_ready()
     }
 
-    fn call(&self, request: Packet) -> Self::Future {
-        self.inner.call(pipeline::Message::WithoutBody(request))
+    fn call(&self, request: Self::Request) -> Self::Future {
+        self.inner.call(request).map(Self::map_err)
     }
 }
 
-impl fmt::Debug for Client {
+impl<Req, Resp, E> Client<Req, Resp, E> {
+    fn new(tcp: TcpStream, handle: &reactor::Handle) -> Self
+        where Req: Serialize + Send + 'static,
+              Resp: Deserialize + Send + 'static,
+              E: Deserialize + Send + 'static
+    {
+        Client {
+            inner: multiplex::connect(Framed::new(tcp), handle),
+        }
+    }
+
+    fn map_err(resp: WireResponse<Resp, E>) -> Result<Resp, ::Error<E>> {
+        resp.map(|r| r.map_err(::Error::from))
+            .map_err(::Error::ClientDeserialize)
```

Comment:
`E` is not guaranteed to be `SerializeError`. `E` is just `E`, but `map_err` is being called on `WireResponse`, whose error type is `DeserializeError`:

```rust
type WireResponse<Resp, E> = Result<Result<Resp, WireError<E>>, DeserializeError>;
```